File: src/altair_cpp_tests/src/altair_cpp_tests/custom_msg_test_node.cpp
#include "altair_cpp_tests/custom_msg_test_node.hpp"



altair::CustomMsgTestNode::CustomMsgTestNode() : rclcpp::Node("CustomMsgTestNode"){
    this->declare_parameter("name", rclcpp::PARAMETER_STRING);
    this->declare_parameter("age", rclcpp::PARAMETER_INTEGER);
    this->declare_parameter("score", rclcpp::PARAMETER_DOUBLE);
    this->declare_parameter("pub_topic", rclcpp::PARAMETER_STRING);
    this->declare_parameter("sub_topic", rclcpp::PARAMETER_STRING);

    this->name      = this->get_parameter("name").as_string();
    this->age       = this->get_parameter("age").as_int();
    this->score     = this->get_parameter("score").as_double();
    this->pub_topic = this->get_parameter("pub_topic").as_string();
    this->sub_topic = this->get_parameter("sub_topic").as_string();

    this->subscriber = this->create_subscription<altair_interfaces::msg::TestMessage>(
        this->sub_topic.c_str(),
        1000,
        std::bind(&altair::CustomMsgTestNode::subCallback, this, _1)
    );

    this->publisher = this->create_publisher<altair_interfaces::msg::TestMessage>(
        this->pub_topic.c_str(),
        1000
    );

    this->pub_timer = this->create_wall_timer(
        1s,
        std::bind(&altair::CustomMsgTestNode::pubTimerCallback, this)
    );
}



altair::CustomMsgTestNode::~CustomMsgTestNode() {
    RCLCPP_INFO(this->get_logger(), "CustomMsgTestNode destroyed");
}



void altair::CustomMsgTestNode::subCallback(const altair_interfaces::msg::TestMessage &msg) {
    RCLCPP_INFO(this->get_logger(), "[%s]: name = %s, age = %d, score = %lf", this->sub_topic.c_str(), msg.name.c_str(), msg.age, msg.score);
}



void altair::CustomMsgTestNode::pubTimerCallback() {
    auto msg = altair_interfaces::msg::TestMessage();
    msg.name    = this->name;
    msg.age     = this->age;
    msg.score   = this->score;
    this->publisher->publish(msg);
    RCLCPP_INFO(this->get_logger(), "message sent to %s", this->pub_topic.c_str());
}

File: src/altair_py_tests/altair_py_tests/modules/custom_msg_test_node.py
import rclpy
from rclpy.node import Node
import altair_interfaces.msg as altair_interfaces



class CustomMsgTestNode(Node):

    def __init__(self) -> None:
        super().__init__('CustomMsgTestNode')

        self.declare_parameter('name', rclpy.Parameter.Type.STRING)
        self.declare_parameter('age', rclpy.Parameter.Type.INTEGER)
        self.declare_parameter('score', rclpy.Parameter.Type.DOUBLE)
        self.declare_parameter('pub_topic', rclpy.Parameter.Type.STRING)
        self.declare_parameter('sub_topic', rclpy.Parameter.Type.STRING)

        self.name       = self.get_parameter('name').value
        self.age        = self.get_parameter('age').value
        self.score      = self.get_parameter('score').value
        self.pub_topic  = self.get_parameter('pub_topic').value
        self.sub_topic  = self.get_parameter('sub_topic').value

        self.subscriber = self.create_subscription(
            msg_type    = altair_interfaces.TestMessage,
            topic       = self.sub_topic,
            callback    = self.subCallback,
            qos_profile = 1000
        )

        self.publisher = self.create_publisher(
            msg_type    = altair_interfaces.TestMessage,
            topic       = self.pub_topic,
            qos_profile = 1000
        )

        self.pub_timer = self.create_timer(1., self.pubTimerCallback)



    def subCallback(self, msg:altair_interfaces.TestMessage) -> None:
        self.get_logger().info(f'[{self.sub_topic}]: name = {msg.name}, age = {msg.age}, score = {msg.score}')



    def pubTimerCallback(self) -> None:
        msg         = altair_interfaces.TestMessage()
        msg.name    = self.name
        msg.age     = self.age
        msg.score   = self.score
        self.publisher.publish(msg)
        
        self.get_logger().info(f'message sent to {self.pub_topic}')